### ISSUE TITLE
Feature request - easy scroll for patches

### ISSUE THREAD

**Issue by proski** [jonas/tig#381]
There should be a way to scroll patches up and down. k and j move the cursor, so they may require several key presses before the view shifts. Also, they only work is the patch window is focused. I think that the windows is an excessive feature and many users are not aware (or don't want to be aware) of the window focus.

Shift-Up and Shift-Down move by the page, which is too much. I want fine control. I want important lines to be in the center of the screen. Shift-Up and Shift-Down are also focus-sensitive.

I suggest that we define keys that always work on the patch window. Let's use the Ctrl modifier for them. There are standard key sequences in xterm. Many other terminals support all of some of those sequences (or can be patched to support them).

Ctrl-Up \e[1;5A
Ctrl-Down \e[1;5B
Ctrl-Left \e[1;5D
Ctrl-Right \e[1;5C

Ctrl-Home \e[1;5H
Ctrl-End \e[1;5F
Ctrl-PgUp \e[5;5~
Ctrl-PgDown \e[6;5~

**Comment by jonas** [comment]
Did you try `scroll-line-up` and `scroll-line-down`?

By default they have the following [bindings](https://github.com/jonas/tig/blob/master/tigrc#L228):
```
bind generic    <Ins>   scroll-line-up
bind generic    <C-Y>   scroll-line-up
bind generic    <Del>   scroll-line-down
bind generic    <C-E>   scroll-line-down
```
Unfortunately ncurses does not support bindings such as `Ctrl-Down`, the key must be one of `a` through to `f` (with certain exceptions). In the spirit of Vim, you could add `Ctrl-j` and `Ctrl-k` bindings.

**Comment by proski** [comment]
Thank you! I see they work in the current window. It's funny, I've been using tig for over a year, but I only discovered today that windows in tig have focus. It's not shown clearly and perhaps it shouldn't. Amazing - my mental model of tig was so different from what it is.

I just found that Up and Down always work on the commits, but Left and Right follow the focus. So inconsistent.

The problem about non-standard keys is that they are easy to forget. It's not a big deal to tig developers, but it's a bigger deal to occasional users. Ctrl-Y is actually "remove a line" in some editors. Ctrl-E is "put cursor to the end of line.

"Hold Ctrl to control the lower window" would be easier to remember.

I know Midnight Commander can parse any escape sequences. I actually wrote some of that code many years ago, and I also wrote some excessive code that caused problems to users (reading X events). Perhaps the good part could be extracted, although it may be easier to write it from scratch.

**Comment by jonas** [comment]
You are not the first one to be confused by this magic handling of view focus.

It is inspired by the news reader `slrn` but I might have "messed" it up along the way. The idea is that you can easily navigate through the history (main view) and individual patches (diff view) using only `Enter` and `Down` (and `Up`). `Down`/`Up` navigates to the next or previous patch in the history and `Enter` either opens the patch view or scrolls it if it is already open. The last part is probably what is surprising. This works for all views that opens a second view in split mode (main/diff, refs/main, blame/diff, tree/blob, grep/blob). The code in question can be found here:

 * https://github.com/jonas/tig/blob/master/src/tig.c#L230

Now, I agree that it is inconsistent when looking at other programs and I am open to changing it in the default configuration if this is deemed necessary. As a start it should probably be documented better. Furthermore, it is quite easy to change this behaviour if you don't like it by changing the bindings for `Down` and `Up` to no invoke the `next` and `previous` actions. For example you can change them to move the cursor instead:

```
bind generic    <Down>  move-down
bind generic    <Up>     move-up
```

Anyway, the initial idea was that `Down` and `Up` are special and is treated as such.

**Comment by proski** [comment]
I agree, documenting the existing behavior would be helpful.

I'll try to configure the bindings that are intuitive, easy to remember and easy to press. In the split view, we want to manipulate the diff and in the same time have keys to navigate patches back and forth. Perhaps `<` and `>` or `[` and `]` would be a good choice.

**Comment by blueyed** [comment]
As a side-note / question: is it possible to jump to the prev/next file in the diff view?

I know that you can jump to a file's diff from the header with `Enter` (which is useful), but am looking for a way similar to `J`/`K`, which navigates between next/prev patches/commits.

**Comment by jonas** [comment]
There's no default binding. You can bind searches to keys, so something like the following in your `~/.tigrc` will do what you want:
```
bind stage <Ctrl-J> :/^diff --
bind stage <Ctrl-K> :?^diff --

bind diff <Ctrl-J> :/^diff --
bind diff <Ctrl-K> :?^diff --
```
... or you can add those bindings in the `generic` keymap.

**Comment by jonas** [comment]
Adding these bindings to `contrib/vim.tigrc` might be useful. See #387 ...

@vivien what do you think?

**Comment by vivien** [comment]
Indeed! Never thought about binding this, really useful! Since `j`, `k` are bound in the default config, should we add this in the main `tigrc` file instead?
